### ISSUE TITLE
Define required testing for new cams

### ISSUE THREAD

**Issue by reticulatedpines** [reticulatedpines/magiclantern_simplified#153]
Similar to issue #152 - what testing do we require for newly supported cams before we can release a build for them?

Currently it is this list:
200D
5D4
6D2
750D 77D 7D2
80D

There is no qemu automation for these cams.  Selftest module is not compatible with these cams and is dangerous to run.  It tries to test things that rely on assumptions about D45 behaviour, e.g. attempting to over allocate from SRM system (safe error on D45, breaks SRM subsystem on D678), or use of null pointers.

Possibly, manual testing of the limited feature set is adequate for an initial release.  I did make a modernised framework for qemu testing, and it's nicer in many ways than the old system...  but it also is far from finished.  Currently it does very little, though it would allow creating new tests.

This ticket is for discussion, leading to a well described testing plan we can follow for all new cams.

## Draft 01

Absolute minimum:
- Cam gets past bootloader in qemu, with ML active
- Dev must test if ML menus work on phys cam
- Dev confirms phys cam feels non-dangerous with ML active

Sensible minimum:
- Above, plus:
- Dev confirms all enabled ML features work as expected.  
  This means testing every single menu option.  
  If it's visible and a user can select it, it must be tested.  
  No 1300D raw video scenarios allowed.

Achievable with effort:
- Above, plus:
- Normal user confirms all features work.  
  User not given a detailed plan, we trust they test sensibly.

Desirable and perhaps possible:
- User works from a checklist we provide to test all features

Deemed impractical for this release, but acknowledged as desireable:
- Broad automated qemu testing

**Comment by reticulatedpines** [comment]
Edited with draft 01.  I propose we do "sensible minimum".  This has no user testing, only dev testing.  We know devs have the cams, and devs will test much better than the average user who doesn't have a test plan.

If anyone thinks we must do "achievable with effort", we can try to find users from Discord and forum.  If we can't find those, we would fail the test plan as described.  What should we do in that case?  If we'd release for that cam anyway, do we really need the user testing?  (I agree it's a "nice to have", but this is my main argument for recommending "sensible minimum")

If anyone thinks we need to do more than that, they will have to create the test checklist for users, and manage the process for collecting test results.

If there is no comment by end of 2025-03-16, we will be doing "sensible minimum".  If any cams don't pass Dev testing, no release for those cams.

**Comment by WalterSchulz** [comment]
ML for new cams comes without FIR bootflag enabler. 
I think we agree UART access to set bootflag should not be recommended as standard installation method.
Therefore petabyte's MLinstall seems to be a way to deal with it. And the only one I'm aware. Are there others?

Required testing definition should include
- Doc describing handling MLinstall for all supported OS. List of incompatible/unsupported OS versions (Linux, macOS, Windows).
- testing enabling/disabling bootflag for new cam with MLinstall. 

Problematic (IMO): macOS is common among potential users. But not among us three (kitor, names_are_hard, myself).
How to handle macOS support?

**Comment by WalterSchulz** [comment]
For discussion:

Cam testing should include connectivity tests in the broadest sense. There are essential ones (lens, ...) and non-essential ones (wireless NFC, BT, ...). Disturbing or disabling essential ones cannot be tolerated and such a build should not be published. All affected/detected connectivity issues should be listed. Untested connectivity features should be listed.

- Lens with electronic contacts should work the same way with/without  ML for photo, video and streaming.
-- Aperture
-- Focus 
-- Stabilizer (if any)
-- (non-essential?) "Retract lens on power off" for supporting cameras-lens combinations.

- Hot shoe
-- External flashes
--- ETTL
--- Native Canon, Third party
--- X-Contact
--- Remote flash control via IR/RT.
-- EVF 
-- (non-essential?) Remote cam control/trigger via RT
-- (non-essential?) Other devices on hot shoe like Canon's GP-E2 (GPS) or Tascam audio device.

- HDMI 
-- HDMI out should work with/without ML. Basic connectivity test with AVR/TV set/computer monitor should be sufficient.
-- (non-essential) Control via HDMI (External -> Cam menu)

- USB
-- Accessing cam via USB and using features like EOS Utility, EOS Webcam Utillity, gphoto2
-- non-essential: Wireless File Transfer (WFT) devices. 

- A/V out (if any. 1xxxD, 2000D, 4000d, 500D, 50D. ... don't)
-- non-essential: connecting to analog TV devices.
-- ML audio functions (headphone monitoring)

- Mic in
-- external microphones working with/without ML.

- Mic (internal)

- Remote trigger via IR/BT

- WLAN

- BT

- NFC

- Extension system terminal
(Some new cams use it for WFT/battery grip)

-

**Comment by reticulatedpines** [comment]
Install options have been discussed elsewhere.  Adding here: if we want something like MLInstall, longer term it should be cross-platform (QT?  pyqt?).  If we want this to be part of the instructions for ML install process, it wants to be something ML team can fix future problems (so, probably open source?  But the real requirement is that we can maintain it).

**Comment by reticulatedpines** [comment]
Re testing and what is "essential".  In general I like the tests you're proposing.  At this early stage of definition, I might group most of the points into "normal cam functionality is not impaired by ML".

There is very little time left for this release.  Do you think it is possible to test everything you list as essential, *and* fix or disable anything requiring changes?  If a test is essential, this implies it is also essential to get the test passing before a release occurs.  We also didn't do these tests regularly (if at all!) on old cams, we relied on users to notice and report problems.

I can see an argument that the old cams were de facto tested in this way and the new cams haven't had enough user hours.  Is that the kind of thing you're thinking?  These tests would then be something like "essential at least once"?

Is it essential that a new cam have as good support as you're suggesting, or is it expected by users that newly supported models won't be as smooth as older ones?  I would have guessed the latter, but I don't really know, you have a longer memory of the ML community.

If you can convince me this level of testing really is essential, I'll likely have to push the release back.  Annoying, but not the end of the world.
